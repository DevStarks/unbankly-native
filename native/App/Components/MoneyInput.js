import React, { Component } from 'react'
import PropTypes from 'prop-types'
// import { View } from 'react-native'
import TextInput from './TextInput'
// import styles from './Styles/MoneyInput'
import ExamplesRegistry from '../Services/ExamplesRegistry'

// Ignore in coverage report
/* istanbul ignore next */

ExamplesRegistry.addComponentExample('MoneyInput', () =>
  <MoneyInput />
)

class MoneyInput extends Component {
  // For a complete list of props
  // https://facebook.github.io/react-native/docs/textinput.html#props

  static propTypes = {
    onChangeText: PropTypes.func.isRequired,
    value: PropTypes.string
  }

  static defaultProps = {
    placeholder: '$0.00'
  }

  constructor (props) {
    super(props)

    this.state = {
      value: props.value
    }
  }

  isValidCurrency (str) {
    // matches numeric str beginning with $ and allows commas and 2 decimals
    return str.match(/^\$([0-9]{1,3},([0-9]{3},)*[0-9]{3}|[0-9]+)?(.[0-9]?[0-9]?)?$/)
  }

  currencyToNumber (currencyValue) {
    return currencyValue.replace(/[$,]/g, '')
  }

  onChangeText (value) {
    if (this.isValidCurrency(value)) {
      this.setState({ value })
      this.props.onChangeText(this.currencyToNumber(value))
    }
  }

  onFocus () {
    if (!this.state.value) {
      this.setState({ value: '$' })
    }
  }

  onBlur () {
    if (this.state.value === '$') {
      this.setState({ value: null })
    }
  }

  render () {
    return (
      <TextInput
        {...this.props}
        value={this.state.value}
        onFocus={this.onFocus.bind(this)}
        onBlur={this.onBlur.bind(this)}
        style={this.props.style}
        onChangeText={this.onChangeText.bind(this)}
        placeholder={this.props.placeholder} />
    )
  }
}

export default MoneyInput
