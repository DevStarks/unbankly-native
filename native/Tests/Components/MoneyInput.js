import React from 'react'
import MoneyInput from '../../App/Components/MoneyInput'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import renderer from 'react-test-renderer'

test('MoneyInput component renders correctly', () => {
  const tree = renderer.create(<MoneyInput onChangeText={() => {}} placeholder='$1.00' />).toJSON()
  expect(tree).toMatchSnapshot()
})
// it calls onnChangeText without punctuation
// it doesn't call onChangeText with invald input
describe('onChangeText', () => {
  test('calls props.onChangeText', () => {
    const onChangeText = sinon.fake()
    const wrapper = shallow(<MoneyInput onChangeText={onChangeText} />)

    wrapper.simulate('changeText', '$1')

    expect(onChangeText.called).toBe(true)
  })

  test('passes number with commas/$ removed to props.onChangeText', () => {
    const onChangeText = sinon.fake()
    const wrapper = shallow(<MoneyInput onChangeText={onChangeText} />)

    wrapper.simulate('changeText', '$1,211.12')

    expect(onChangeText.calledWith('1211.12')).toBe(true)
  })

  test('doesnt call props.onChangeText when input is non currency', () => {
    const onChangeText = sinon.fake()
    const wrapper = shallow(<MoneyInput onChangeText={onChangeText} />)

    wrapper.simulate('changeText', 'bad')

    expect(onChangeText.called).toBe(false)
  })
})

test('onFocus sets state.value', () => {
  const wrapper = shallow(<MoneyInput onChangeText={() => {}} />)

  wrapper.simulate('focus')

  expect(wrapper.state().value).toBe('$')
})

test('onBlur nulls value', () => {
  const wrapper = shallow(<MoneyInput onChangeText={() => {}} />)

  wrapper.simulate('focus')
  wrapper.simulate('blur')

  expect(wrapper.state().value).toBe(null)
})
