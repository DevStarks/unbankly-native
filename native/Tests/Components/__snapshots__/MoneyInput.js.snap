// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneyInput component renders correctly 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "height": 60,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="$1.00"
    style={
      Object {
        "borderColor": "#E7E7E7",
        "borderWidth": 1,
        "color": "#637381",
        "fontFamily": "AmaticSC-Bold",
        "fontSize": 20,
        "height": 60,
        "paddingBottom": 16,
        "paddingLeft": 5,
        "paddingTop": 16,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Object {
        "backgroundColor": "#ECC925",
        "bottom": 4,
        "height": 1,
      }
    }
  />
</View>
`;
